File: java_client/src/main/java/com/cisc468share/storage/VaultStore.java
package com.cisc468share.storage;

import com.cisc468share.crypto.Vault;
import com.cisc468share.protocol.CanonicalJson;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Manages persistent vault storage.
 */
public class VaultStore {
    private final Path vaultPath;
    private final String password;
    private final Path indexPath;
    private final ObjectMapper mapper = CanonicalJson.mapper();

    public VaultStore(Path vaultPath, String password) {
        this.vaultPath = vaultPath;
        this.password = password;
        this.indexPath = vaultPath.resolve("index.json");
        try {
            Files.createDirectories(vaultPath);
        } catch (Exception e) {
            throw new RuntimeException("Failed to create vault directory", e);
        }
    }
    
    /**
     * Store an encrypted file.
     * 
     * @param fileId The file identifier
     * @param encryptedData The encrypted file data
     */
    public void storeEncryptedFile(String fileId, byte[] encryptedData) {
        try {
            Files.write(encryptedPath(fileId), encryptedData);
            updateIndex(fileId, fileId + ".enc");
        } catch (Exception e) {
            throw new RuntimeException("Failed to store encrypted file", e);
        }
    }
    
    /**
     * Retrieve an encrypted file.
     * 
     * @param fileId The file identifier
     * @return The encrypted file data
     */
    public byte[] retrieveEncryptedFile(String fileId) {
        try {
            return Files.readAllBytes(encryptedPath(fileId));
        } catch (Exception e) {
            throw new RuntimeException("Failed to read encrypted file", e);
        }
    }
    
    /**
     * Delete a file from the vault.
     * 
     * @param fileId The file identifier
     */
    public void deleteFile(String fileId) {
        try {
            Files.deleteIfExists(encryptedPath(fileId));
            Map<String, String> index = loadIndex();
            index.remove(fileId);
            saveIndex(index);
        } catch (Exception e) {
            throw new RuntimeException("Failed to delete vault file", e);
        }
    }
    
    /**
     * List all files in the vault.
     * 
     * @return List of file IDs
     */
    public java.util.List<String> listFiles() {
        return new ArrayList<>(loadIndex().keySet());
    }

    public void storeFile(String fileId, byte[] data) {
        try {
            storeEncryptedFile(fileId, Vault.encryptVault(password, data));
        } catch (Exception e) {
            throw new RuntimeException("Failed to encrypt vault file", e);
        }
    }

    public byte[] getFile(String fileId) {
        try {
            return Vault.decryptVault(password, retrieveEncryptedFile(fileId));
        } catch (Exception e) {
            throw new RuntimeException("Failed to decrypt vault file", e);
        }
    }

    private Path encryptedPath(String fileId) {
        return vaultPath.resolve(fileId + ".enc");
    }

    private Map<String, String> loadIndex() {
        if (!Files.exists(indexPath)) {
            return new LinkedHashMap<>();
        }

        try {
            return mapper.readValue(indexPath.toFile(), new TypeReference<Map<String, String>>() {});
        } catch (Exception e) {
            return new LinkedHashMap<>();
        }
    }

    private void saveIndex(Map<String, String> index) {
        try {
            mapper.writerWithDefaultPrettyPrinter().writeValue(indexPath.toFile(), index);
        } catch (Exception e) {
            throw new RuntimeException("Failed to write vault index", e);
        }
    }

    private void updateIndex(String fileId, String encryptedName) {
        Map<String, String> index = loadIndex();
        index.put(fileId, encryptedName);
        saveIndex(index);
    }
}

File: python_client/storage/vault_store.py
"""
Encrypted file vault storage.

Files stored here are encrypted with AES-256-GCM using a key derived from
a user password via PBKDF2-HMAC-SHA-256 (200,000 iterations).
An attacker who steals the device cannot read vault files without the password.
"""
import json
import os
from pathlib import Path
from crypto.vault import encrypt_vault, decrypt_vault


class VaultStore:
    """
    Stores files encrypted at rest using a password-derived key.

    Each file is stored as: <vault_dir>/<filename>.enc
    The format is: salt(16) || nonce(12) || AES-GCM-ciphertext
    """

    def __init__(self, vault_dir: Path, password: str = "default-vault-password"):
        self.vault_dir = Path(vault_dir)
        self.password = password
        self.vault_dir.mkdir(parents=True, exist_ok=True)
        # Index maps filename -> encrypted filename on disk
        self._index_path = self.vault_dir / "index.json"

    def store_file(self, filename: str, data: bytes):
        """
        Encrypt and store a file in the vault.

        Args:
            filename: The logical filename (as the peer shared it)
            data: Raw file bytes
        """
        encrypted = encrypt_vault(self.password, data)
        enc_path = self.vault_dir / (filename + ".enc")
        enc_path.write_bytes(encrypted)
        self._update_index(filename, filename + ".enc")

    def get_file(self, filename: str) -> bytes:
        """
        Retrieve and decrypt a file from the vault.

        Args:
            filename: The logical filename

        Returns:
            bytes: Decrypted file data

        Raises:
            FileNotFoundError: If the file is not in the vault
            Exception: If decryption fails (wrong password or tampered data)
        """
        enc_path = self.vault_dir / (filename + ".enc")
        if not enc_path.exists():
            raise FileNotFoundError(f"File '{filename}' not found in vault")
        encrypted = enc_path.read_bytes()
        return decrypt_vault(self.password, encrypted)

    def list_files(self) -> list:
        """Return list of filenames stored in the vault."""
        index = self._load_index()
        return list(index.keys())

    def has_file(self, filename: str) -> bool:
        """Check if a file exists in the vault."""
        return (self.vault_dir / (filename + ".enc")).exists()

    def delete_file(self, filename: str):
        """Delete a file from the vault."""
        enc_path = self.vault_dir / (filename + ".enc")
        if enc_path.exists():
            enc_path.unlink()
        index = self._load_index()
        index.pop(filename, None)
        self._save_index(index)

    def _load_index(self) -> dict:
        if not self._index_path.exists():
            return {}
        return json.loads(self._index_path.read_text())

    def _save_index(self, index: dict):
        self._index_path.write_text(json.dumps(index, indent=2))

    def _update_index(self, filename: str, enc_filename: str):
        index = self._load_index()
        index[filename] = enc_filename
        self._save_index(index)
